runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+		join information_schema.columns c on t.table_schema = c.table_schema and t.table_name = c.table_name
+		group by 1, 2, 3, 4
+		order by 1, 2, 3, 4
+	`)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+func (d *db) attachDBs(ctx context.Context, db *sqlx.DB, path string, read bool) error {
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	var views []string
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		// NOTE :: we always look at the write version
+		// Tables in read path are removed after getting a new handle
+		// So we need to always look at the write version to ensure we do not reattach dropped tables
+		version, exist, _ := d.writeTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		versionPath := filepath.Join(path, entry.Name(), version)
+
+		// read meta file
+		f, err := os.ReadFile(filepath.Join(versionPath, "meta.json"))
+		if err != nil {
+			_ = os.RemoveAll(versionPath)
+			d.logger.Warn("error in reading meta file", slog.String("table", entry.Name()), slog.Any("error", err))
+			return err
+		}
+		var meta meta
+		err = json.Unmarshal(f, &meta)
+		if err != nil {
+			_ = os.RemoveAll(versionPath)
+			d.logger.Warn("error in unmarshalling meta file", slog.String("table", entry.Name()), slog.Any("error", err))
+			return err
+		}
+
+		if meta.ViewSQL != "" {
+			// table is a view
+			views = append(views, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s\n)", safeSQLName(entry.Name()), meta.ViewSQL))
+			continue
+		}
+		switch meta.Format {
+		case BackupFormatDB:
+			dbName := dbName(entry.Name())
+			var readMode string
+			if read {
+				readMode = " (READ_ONLY)"
+			}
+			_, err := db.ExecContext(ctx, fmt.Sprintf("ATTACH %s AS %s %s", safeSQLString(filepath.Join(versionPath, "data.db")), safeSQLName(dbName), readMode))
+			if err != nil {
+				d.logger.Error("error in attaching db", slog.String("table", entry.Name()), slog.Any("error", err))
+				_ = os.RemoveAll(filepath.Join(path, entry.Name()))
+				return err
+			}
+
+			_, err = db.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS SELECT * FROM %s.default", safeSQLName(entry.Name()), safeSQLName(dbName)))
+			if err != nil {
+				return err
+			}
+		case BackupFormatParquet:
+			panic("unimplemented")
+		default:
+			return fmt.Errorf("unknown backup format %q", meta.Format)
+		}
+	}
+	// create views after attaching all the DBs since views can depend on other tables
+	for _, view := range views {
+		_, err := db.ExecContext(ctx, view)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *db) readTableVersion(name string) (string, bool, error) {
+	return tableVersion(d.readPath, name)
+}
+
+func (d *db) writeTableVersion(name string) (string, bool, error) {
+	return tableVersion(d.writePath, name)
+}
+
+func execIncrementalInsert(ctx context.Context, conn *sqlx.Conn, safeTableName, query string, opts *InsertTableOptions) error {
```

Comment:
See earlier comment – do we really need incremental insert logic implemented in this package? It would be nice if the executor could provide the logic, and this package just takes care of creating a new handle for the new DB file (and expects a callback to create/mutate the `default` table).